bundles/org.openhab.binding.openthermgateway/src/main/java/org/openhab/binding/openthermgateway/internal/OpenThermGatewaySocketConnector.java

```diff
@@ -0,0 +1,215 @@
+                writer = wrt;
+
+                sendCommand(GatewayCommand.parse(GatewayCommandCode.PrintReport, "A"));
+                // Set the OTGW to report every message it receives and transmits
+                sendCommand(GatewayCommand.parse(GatewayCommandCode.PrintSummary, "0"));
+
+                while (!stopping && !Thread.currentThread().isInterrupted()) {
+                    @Nullable
+                    String message = reader.readLine();
+
+                    if (message != null) {
+                        handleMessage(message);
+                    } else {
+                        logger.debug("Connection closed by OpenTherm Gateway");
+                        break;
+                    }
+                }
+
+                logger.debug("Stopping OpenThermGatewaySocketConnector");
+            } finally {
+                connected = false;
+
+                logger.debug("OpenThermGatewaySocketConnector disconnected");
+                callback.disconnected();
+            }
+        } catch (IOException ex) {
+            logger.warn("Unable to connect to the OpenTherm Gateway.", ex);
+        }
+    }
+
+    @Override
+    public void stop() {
+        logger.debug("Stopping OpenThermGatewaySocketConnector");
+        stopping = true;
+    }
+
+    @Override
+    public boolean isConnected() {
+        return connected;
+    }
+
+    @Override
+    public void sendCommand(GatewayCommand command) {
+        @Nullable
+        PrintWriter wrtr = writer;
+
+        String msg = command.toFullString();
+
+        pendingCommands.put(command.getCode(),
+                new AbstractMap.SimpleImmutableEntry<>(System.currentTimeMillis(), command));
+
+        if (connected) {
+            logger.debug("Sending message: {}", msg);
+            if (wrtr != null) {
+                wrtr.printf("%s\r\n", msg);
+                wrtr.flush();
+            }
+        } else {
+            logger.debug("Unable to send message: {}. OpenThermGatewaySocketConnector is not connected.", msg);
+        }
+    }
+
+    private void handleMessage(String message) {
+        if (message.length() > 2 && message.charAt(2) == ':') {
+            String code = message.substring(0, 2);
+            String value = message.substring(3);
+
+            logger.debug("Received command confirmation: {}: {}", code, value);
+            pendingCommands.remove(code);
+            return;
+        }
+
+        for (Entry<Long, GatewayCommand> timeAndCommand : pendingCommands.values()) {
+            if (System.currentTimeMillis() > timeAndCommand.getKey() + COMMAND_RESPONSE_TIME_MILLISECONDS) {
+                logger.debug("Resending command: {}", timeAndCommand.getValue().toFullString());
+                sendCommand(timeAndCommand.getValue());
+            } else if (System.currentTimeMillis() > timeAndCommand.getKey() + COMMAND_TIMEOUT_MILLISECONDS) {
+                pendingCommands.remove(timeAndCommand.getValue().getCode());
+            }
```

Comment:
This code will never get called because you compare the time against the smallest value first.

Reply:
Changed the check